### PR TITLE
docs: rfc development process

### PR DESCRIPTION
**Pull request by Kononnable**
Should clarify some things about what to do with tasks during a sprint.

Some of it is proposals, others are writing down what we're already using. So as usual everything is up for discussion.

Rendered: https://github.com/epiphany-platform/CommonDataLayer/blob/rfc-dev-process/docs/rfc/NAF-Development_process.md

I did not make changes to RFC lists etc, those things are planned for a lot of changes, will add once PR is closer to merging.

### CODE REVIEW

File: docs/rfc/NAF-Development_process.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+BLOCKED - a person responsible for checking if blockade can be removed, most often a person to which issue was assigned before entering BLOCKED 
+
+TODO - no assign, exceptions: bug issues getting back from GENERAL TESTS state(person implementing the issue), on planning we may decide that specific task should be done by a specific person(on per task basis)
+
+IN PROGRESS - person actively working on the issue
+
+CODE REVIEW - a person who implemented the issue and the main reviewer(picked automatically) - it's often for reviewers to have some comments that have to be addressed by the author
+
+ACCEPTANCE TESTS - author of RFC describing the feature, can be changed to another person
```
Comment (frondeus) [inline docs/rfc/NAF-Development_process.md:110]: ACCEPTANCE TEST is not for RFC

Comment (Kononnable) [reply docs/rfc/NAF-Development_process.md:110]: Most feature changes/additions are based on RFC. Acceptance tests are not part of RFC process, however, It might be good to let the creator of RFC that change is based on to check for covering all the cases after the change is implemented. In most scenarios, this should be a person knowing what exactly he/she proposed, what use cases they were trying to solve, etc.
I'm proposing RFC creator here to make sure that there is no miscommunication on what RFC creator and the team know about this change solving user cases. RFC should cover all the use cases, but sometimes things go missing - this is just a fallback check.

File: docs/rfc/NAF-Development_process.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+| ANY         | BLOCKED     | Moving task to next state is blocked by something, not dependant on the team(e.g. other team decision/reply) or on another issue |
+|BLOCKED      | _PREVIOUS_  | Blocking reason was resolved. If resolve reason changes the task drastically new state should be TO DO or IN PROGRESS |
+| ANY         | TODO        | During the work on the task we decided that it should be redefined/we need to take care of the problem in a different way, should not happen often(if at all), if we don't know the solution(needs RFC) task should go to BLOCKED instead |
+| TODO        | IN PROGRESS |  Someone started working on the issue |
+| IN PROGRESS | CODE REVIEW | Implementation work/Design is done, acceptance criteria are achieved |
+### Feature
+| BEFORE           | AFTER           | CHANGE CONDITION     |
+|------------------|-----------------|----------------------|
+| CODE REVIEW      | ACCEPTANCE TESTS| Code quality is fine |
+| ACCEPTANCE TESTS | GENERAL TESTS   | Functionality works as defined in documentation |
```
Comment (esavier) [inline docs/rfc/NAF-Development_process.md:67]: we usually do not have up to date docs, you mean RFC? also some things do not have RFCs, then what

File: docs/rfc/NAF-Development_process.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+|BLOCKED      | _PREVIOUS_  | Blocking reason was resolved. If resolve reason changes the task drastically new state should be TO DO or IN PROGRESS |
+| ANY         | TODO        | During the work on the task we decided that it should be redefined/we need to take care of the problem in a different way, should not happen often(if at all), if we don't know the solution(needs RFC) task should go to BLOCKED instead |
+| TODO        | IN PROGRESS |  Someone started working on the issue |
+| IN PROGRESS | CODE REVIEW | Implementation work/Design is done, acceptance criteria are achieved |
+### Feature
+| BEFORE           | AFTER           | CHANGE CONDITION     |
+|------------------|-----------------|----------------------|
+| CODE REVIEW      | ACCEPTANCE TESTS| Code quality is fine |
+| ACCEPTANCE TESTS | GENERAL TESTS   | Functionality works as defined in documentation | 
+| GENERAL TESTS    | DONE            | Functionality works on all supported configurations, corner cases, error handling are covered by tests |
```
Comment (esavier) [inline docs/rfc/NAF-Development_process.md:68]: that requires test driven development, which in turn requires us to first do spec, then tests, then actually code,

File: docs/rfc/NAF-Development_process.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+| ACCEPTANCE TESTS | IN PROGRESS     | Adding acceptance tests reviled that some parts of the functionality are missing/major use case is not working correctly. If that use case wasn't covered by documentation new feature related task should be added instead. |
```
Comment (esavier) [inline docs/rfc/NAF-Development_process.md:69]: again by docs you mean RFC?
ususally it should be done case by case since there are a lot of things that can go wrong in that case,

If usecase was not covered in the docs it means it's either out of scope or rfc is done incorrectly.
how do you know if use case is missing if it's not in the docs? Additionally, i would say it's architect responsibility to check if rfc covers all use cases and define the scope of the feature according to people who requested it.

File: docs/rfc/NAF-Development_process.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+### Bug
+| BEFORE      | AFTER        | CHANGE CONDITION |
+|-------------|--------------|------------------|
+| CODE REVIEW | GENERAL TESTS| Code quality is fine, task should go to the top of the column(prioritizing) |
```
Comment (esavier) [inline docs/rfc/NAF-Development_process.md:74]: its not always on the top, its done so per usecase

File: docs/rfc/NAF-Development_process.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+| GENERAL TESTS | DONE | Functionality works on all supported configurations, corner cases, error handling are covered by tests  |
```
Comment (esavier) [inline docs/rfc/NAF-Development_process.md:75]: its not defined what is configuration, nor we have tests to do that. There is a lot going on in this sentence.
We can not afford to enforce it right now despite how much i would love to.

File: docs/rfc/NAF-Development_process.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+|GENERAL TESTS | TO DO | Bug was not resolved. It's up to the tester to decide if the issue closed(during testing another wrong behavior was observed) or sent back to TO DO(the same wrong behavior still happens). If an issue is sent back person previously implementing the task should be assigned and the issue should go to the top of the TODO list. |
```
Comment (esavier) [inline docs/rfc/NAF-Development_process.md:76]: "...ck person previously implementing the ta..." missing comma i guess

Comment (esavier) [reply docs/rfc/NAF-Development_process.md:76]: well, a lot of missing commas...

Comment (esavier) [inline docs/rfc/NAF-Development_process.md:76]: why person doing it previously should be assigned?

Comment (esavier) [inline docs/rfc/NAF-Development_process.md:76]: Prioritization is another topic and again, should be done in per-case scenario

File: docs/rfc/NAF-Development_process.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+### Bug
+| BEFORE      | AFTER        | CHANGE CONDITION |
+|-------------|--------------|------------------|
+| CODE REVIEW | GENERAL TESTS| Code quality is fine, task should go to the top of the column(prioritizing) |
+| GENERAL TESTS | DONE | Functionality works on all supported configurations, corner cases, error handling are covered by tests  |
+|GENERAL TESTS | TO DO | Bug was not resolved. It's up to the tester to decide if the issue closed(during testing another wrong behavior was observed) or sent back to TO DO(the same wrong behavior still happens). If an issue is sent back person previously implementing the task should be assigned and the issue should go to the top of the TODO list. | 
+### Documentation
+| BEFORE      | AFTER | CHANGE CONDITION              |
+|-------------|-------|-------------------------------|
+| CODE REVIEW | DONE  | Documentation quality is fine |
```
Comment (esavier) [inline docs/rfc/NAF-Development_process.md:80]: can't we just mention that docs are not a subject for testing?

File: docs/rfc/NAF-Development_process.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+### RFC
+| BEFORE      | AFTER | CHANGE CONDITION |
+|-------------|-------|------------------|
+| CODE REVIEW | DONE  | RFC meeting ended with a positive conclusion. Problems mentioned at the meeting were fixed. PR with RFC are merged. |
```
Comment (esavier) [inline docs/rfc/NAF-Development_process.md:85]: what if rfc is a r'n'd and rfc do not have conclusion?

File: docs/rfc/NAF-Development_process.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+### Internal enhancements, chores
+| BEFORE      | AFTER | CHANGE CONDITION |
+|-------------|-------|------------------|
+| CODE REVIEW | DONE  | Code quality is fine, change works correctly(manual testing may be required) |
```
Comment (esavier) [inline docs/rfc/NAF-Development_process.md:90]: what if internal enhancement should go through the testing? i.e. library was changed or something similar?

File: docs/rfc/NAF-Development_process.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+- TODO: default state for new issues, work on the issue resolution has not yet been started 
+- IN PROGRESS: task on which there is an active development
+- CODE REVIEW: implementation is completed, the review should focus mostly on the code, checks its quality; for chores and other internal enhancements it should include testing if the change actually works(if there is something to test manually)
+- ACCEPTANCE TESTS: waiting for automated tests to be written on new/changed functionality. Those tests should focus mostly on covering all of the use cases - check if business logic is implemented properly and if the feature is useful for the user. Corner cases and error handling doesn't have to be tested in this state. Additionally, architecture and general documentation should be written in the process(if not extracted to another task - needs discussion).
+- GENERAL TESTS: waiting for automated and/or manual tests to be done on the functionality - checking for corner cases, proper error handling/propagation, testing in multiple different configurations(e.g. making sure feature works on Kafka, RabbitMQ, and Grpc environments)
```
Comment (esavier) [inline docs/rfc/NAF-Development_process.md:42]: Manual testing have to be performed by developer asignedto the task (code author). manual testing should (but not must) be done again by the reviewer.

File: docs/rfc/NAF-Development_process.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+
+## Definitions of each state:
+- BLOCKED - special state, issues from any state can go directly to this column is blocked by something, when leaving they should go back to the previous state unless a resolution of blockade changed some important details about the task
+- TODO: default state for new issues, work on the issue resolution has not yet been started 
+- IN PROGRESS: task on which there is an active development
+- CODE REVIEW: implementation is completed, the review should focus mostly on the code, checks its quality; for chores and other internal enhancements it should include testing if the change actually works(if there is something to test manually)
+- ACCEPTANCE TESTS: waiting for automated tests to be written on new/changed functionality. Those tests should focus mostly on covering all of the use cases - check if business logic is implemented properly and if the feature is useful for the user. Corner cases and error handling doesn't have to be tested in this state. Additionally, architecture and general documentation should be written in the process(if not extracted to another task - needs discussion).
```
Comment (esavier) [inline docs/rfc/NAF-Development_process.md:41]: those are done automatically, why do we need to for another state of something being done automatically by the github?

File: docs/rfc/NAF-Development_process.md [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+|-------------|-------|------------------|
+| CODE REVIEW | DONE  | RFC meeting ended with a positive conclusion. Problems mentioned at the meeting were fixed. PR with RFC are merged. |
+
+### Internal enhancements, chores
+| BEFORE      | AFTER | CHANGE CONDITION |
+|-------------|-------|------------------|
+| CODE REVIEW | DONE  | Code quality is fine, change works correctly(manual testing may be required) |
+## What is considered an undelivered task in sprint scope
+
+ In general, tasks taken for the sprint should be finished by end of the sprint. However, it's common to define some states which will be treated as valid for sprint end.
```
Comment (esavier) [inline docs/rfc/NAF-Development_process.md:93]: they are not, since they can drop back to todo, its manager's decision if sprint is delivered.

File: docs/rfc/NAF-Development_process.md [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+| CODE REVIEW | DONE  | RFC meeting ended with a positive conclusion. Problems mentioned at the meeting were fixed. PR with RFC are merged. |
+
+### Internal enhancements, chores
+| BEFORE      | AFTER | CHANGE CONDITION |
+|-------------|-------|------------------|
+| CODE REVIEW | DONE  | Code quality is fine, change works correctly(manual testing may be required) |
+## What is considered an undelivered task in sprint scope
+
+ In general, tasks taken for the sprint should be finished by end of the sprint. However, it's common to define some states which will be treated as valid for sprint end. 
+Due to the difference in development vs testing manpower I recommend we don't treat ** feature-related** tasks in the GENERAL TESTING state as not-delivered. Such tasks should be added to the next sprint in the same column. This should ease our planning sessions and make workflow more fluid(as usually there are no new tasks to test at the beginning of the sprint).
```
Comment (esavier) [inline docs/rfc/NAF-Development_process.md:94]: how would you mark something as non feaure-related?

Comment (esavier) [inline docs/rfc/NAF-Development_process.md:94]: > Such tasks should be added to the next sprint in the same column.

again, case by case

File: docs/rfc/NAF-Development_process.md [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+
+> Note: Bugs in the GENERAL TESTING state should mark sprint as undelivered. They can move tasks back on the pipeline - they check if the issue was solved(with feature-related tasks this requirement is met on ACCEPTANCE TESTS state).
```
Comment (esavier) [inline docs/rfc/NAF-Development_process.md:96]: > Bugs in the GENERAL TESTING state should mark sprint as undelivered. They can move tasks back on the pipeline

...and again, case by case, some should, some shouldn't

File: docs/rfc/NAF-Development_process.md [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+
+## Assignees
+We can decide that tasks in specific states should have the correct people assigned to them. Assigning a proper person is a responsibility by of person changing the task state.
+Those are general rules - we may make exceptions on per-issue cases(e.g. someone is out of office for the rest of the sprint, etc.).
```
Comment (esavier) [inline docs/rfc/NAF-Development_process.md:100]: sooooo... those are the rules, but we will still do it, you know, case-by-case :wink:  ?

File: docs/rfc/NAF-Development_process.md [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+
+BLOCKED - a person responsible for checking if blockade can be removed, most often a person to which issue was assigned before entering BLOCKED 
+
+TODO - no assign, exceptions: bug issues getting back from GENERAL TESTS state(person implementing the issue), on planning we may decide that specific task should be done by a specific person(on per task basis)
+
+IN PROGRESS - person actively working on the issue
```
Comment (esavier) [inline docs/rfc/NAF-Development_process.md:106]: persons involved in the issue, you do not have to be working on it full time, or in cases of per-programming it will be weird.

File: docs/rfc/NAF-Development_process.md [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+
+> Note: Bugs in the GENERAL TESTING state should mark sprint as undelivered. They can move tasks back on the pipeline - they check if the issue was solved(with feature-related tasks this requirement is met on ACCEPTANCE TESTS state).
+
+## Assignees
+We can decide that tasks in specific states should have the correct people assigned to them. Assigning a proper person is a responsibility by of person changing the task state.
+Those are general rules - we may make exceptions on per-issue cases(e.g. someone is out of office for the rest of the sprint, etc.).
+
+BLOCKED - a person responsible for checking if blockade can be removed, most often a person to which issue was assigned before entering BLOCKED 
+
+TODO - no assign, exceptions: bug issues getting back from GENERAL TESTS state(person implementing the issue), on planning we may decide that specific task should be done by a specific person(on per task basis)
```
Comment (esavier) [inline docs/rfc/NAF-Development_process.md:104]: >  on planning we may decide that specific task should be done by a specific person(on per task basis)

and how will you mark it?
whats wrong in asking on comms person who is assigned if its free to be taken?

**Review by esavier — CHANGES_REQUESTED** [review]

**Comment by Kononnable** [comment]
Closing as the idea was completely misunderstood by at least half of the team.  I don't think moving these ideas forward is worth the time and energy investment in the process.
